src/lib.rs

```diff
@@ -0,0 +1,310 @@
+mod baseline;
+mod check_release;
+mod config;
+mod dump;
+mod manifest;
+mod query;
+mod templating;
+mod util;
+
+pub use config::*;
+pub use query::*;
+
+use check_release::run_check_release;
+use trustfall_rustdoc::load_rustdoc;
+
+use std::{collections::HashSet, path::PathBuf};
+
+/// Test a release for semver violations.
+#[derive(Default)]
+pub struct Check {
+    /// Which packages to analyze.
+    scope: Scope,
+    current: Current,
+    baseline: Baseline,
+    log_level: Option<log::Level>,
+}
+
+#[derive(Default)]
+enum Baseline {
+    /// Version from registry to lookup for a baseline. E.g. "1.0.0".
+    Version(String),
+    /// Git revision to lookup for a baseline.
+    Revision(String),
+    /// Directory containing baseline crate source.
+    Root(PathBuf),
+    /// The rustdoc json file to use as a semver baseline.
+    RustDoc(PathBuf),
+    /// Latest version published to the cargo registry.
+    #[default]
+    LatestVersion,
+}
+
+/// Current version of the project to analyze.
+#[derive(Default)]
+enum Current {
+    /// Path to the manifest of the current version of the project.
+    /// It can be a workspace or a single package.
+    Manifest(PathBuf),
+    /// The rustdoc json of the current version of the project.
+    RustDoc(PathBuf),
```

Comment:
I saw that this rustdoc also requires the Baseline rustdoc.
At the moment this is not captured by the type system. How would you deal with this?

Reply:
So we want to create a RustdocSource struct that can be used both by the Baseline and Current rustdoc?

> already generated, just load it from an existing file

```rust
enum RustdocSource {
    RustDoc(PathBuf)
}
```

> generate it based on code at some path, optionally first checking out a gitrev (maybe better split into two cases? what do you think?)



```rust
enum RustdocSource {
    RustDoc(PathBuf),
    Root(PathBuf),
    // (path to root, rev)
    Revision(PathBuf, String)
}
```

> generate it based on a registry version

```rust
enum RustdocSource {
    RustDoc(PathBuf),
    Root(PathBuf),
    // (path to root, rev)
    Revision(PathBuf, String),
    Version(Option<String>),
}
```

This looks similar to the current Baseline enum. Is it valid for `Current`, too?
If I didn't get it, can you give me the structure of the enum you were thinking about?

Reply:
I think it looks reasonable. It currently isn't valid for `Current` in the CLI, but I think it should be in the lib. Not much reason not to: that would make the types more complex while providing less functionality.

Small nit: let's call the variant `Rustdoc` instead, since the underlying tool doesn't capitalize the `D` either.

Reply:
> It currently isn't valid for Current in the CLI, but I think it should be in the lib. 

FYI @MarcoIeni, the recent merged chain of PRs (which ended with https://github.com/obi1kenobi/cargo-semver-checks/pull/351) made it so that the `current` crate's rustdoc is generated in the same way as the `baseline`'s rustdoc, so now it's possible to create the `current` rustdoc from registry and from git revision -- your `RustdocSource` now is also fully valid for the `current` crates.

Reply:
And also, this chain of PRs renamed quite a lot of things in the source code, so there might be a big merge conflict, sorry :( I can try to merge main into this branch when I'll have some free time.

Reply:
If you could help with the merge, that would be great @tonowak. Probably the highest impact next thing you can do right now, TBH.

Reply:
Ok, I'll do it in max few days.

Reply:
Done.